src/dataflow/src/source/postgres.rs

```diff
@@ -304,23 +333,21 @@ impl SimpleSource for PostgresSourceReader {
         .await
         .map_err(|e| SourceError::FileIO(e.to_string()))?;
 
-        let client = self
-            .connect_replication()
-            .await
-            .map_err(|e| SourceError::FileIO(e.to_string()))?;
-
         // The initial snapshot has no easy way of retrying it in case of connection failures
-        self.produce_snapshot(&client, &table_info, timestamper)
+        self.produce_snapshot(&table_info, timestamper)
             .await
             .map_err(|e| SourceError::FileIO(e.to_string()))?;
 
-        // TODO(petrosagg): implement retry logic for the replication phase. This is tricky because
-        // in order to do it right we need to specify an LSN that will not skip any data but the
-        // logical decode stream is not in LSN order and it's unclear what LSN we should use
-        self.produce_replication(&client, &table_info, timestamper)
-            .await
-            .map_err(|e| SourceError::FileIO(e.to_string()))?;
+        loop {
+            match self.produce_replication(&table_info, timestamper).await {
+                Ok(_) => log::info!("replication interrupted with no error"),
+                Err(ReplicationError::Recoverable(e)) => {
```

Comment:
Ooh, cool! Should there be a limit to how many times we retry when recoverable errors are thrown?

Reply:
Hm, I wouldn't put a hard limit as in reality there is nothing wrong with an unreachable upstream. It could be that it's offline for a day and materialize is fine with it.

What I think it's missing (I'll add a TODO for it) is implement some sort of exponential backoff. What do you think?

Reply:
That makes sense to me, thanks for clarifying!